src/hooks/useOrders.ts

```diff
@@ -27,19 +28,130 @@ function filterDeletedOrders(orders: AuctionElement[]): AuctionElement[] {
+  const [lastUserAddress, setLastUserAddress] = useSafeState<string | undefined>(undefined)
+
+  const _fetchOrders = useCallback(
+    async (
+      userAddress: string | undefined,
+      networkId: number | undefined,
+      offset: number | undefined,
+      areThereOrders?: boolean,
+    ): Promise<void> => {
+      if (userAddress === undefined || networkId === undefined || offset === undefined) {
+        return
+      }
+
+      await exchangeApi.getOrdersPaginated({ userAddress, networkId, offset }).then(({ orders, nextIndex }) => {
+        if (orders.length) {
+          // Apply filters (remove deleted orders)
+          const filteredOrders = filterDeletedOrders(orders)
+
+          // Store new orders, if any
+          if (offset === 0) {
+            // fresh start/refresh: replace whatever is stored
+            setOrders(filteredOrders)
+          } else if (filteredOrders.length) {
+            // incremental update: append
+            setOrders(curr => curr.concat(filteredOrders))
+          }
+        } else if (offset === 0 && areThereOrders) {
+          // There were orders. We fetched again from the beginning, and now there are none (in the first page). Set to []
+          // Example: switching network or address.
+          setOrders([])
+        }
+
+        // `nextIndex` can be `undefined`, which means there are no more pages
+        if (nextIndex) {
+          // There are more, keep going
+          _fetchOrders(userAddress, networkId, nextIndex)
+        } else {
+          // We are done, remember where we stopped for next time
+          setOffset(offset + orders.length)
+        }
+      })
+    },
+    [],
+  )
+
+  const fetchOrdersAndHandleErrors = useCallback(
+    async (
+      userAddress: string | undefined,
+      networkId: number | undefined,
+      offset: number | undefined,
+      areThereOrders?: boolean,
+    ): Promise<void> => {
+      setIsLoading(true)
+      try {
+        await _fetchOrders(userAddress, networkId, offset, areThereOrders)
+      } catch (e) {
+        console.log('Failed to fetch orders', e)
+        // TODO: inform user
+      } finally {
+        setIsLoading(false)
+      }
+    },
+    [],
+  )
+
+  useEffect(() => {
+    setLastBlockNumber(blockNumber)
+  }, [blockNumber, setLastBlockNumber])
+
+  useEffect(() => {
+    setLastNetworkId(networkId)
+  }, [networkId, setLastNetworkId])
 
   useEffect(() => {
-    userAddress &&
-      networkId &&
-      exchangeApi
-        .getOrders({ userAddress, networkId })
-        .then(filterDeletedOrders)
-        .then(setOrders)
-    // updating list of orders on every block by listening on `blockNumber`
-  }, [networkId, setOrders, userAddress, blockNumber])
-
-  return orders
+    setLastUserAddress(userAddress)
+  }, [userAddress, setLastUserAddress])
+
+  useEffect(() => {
+    // whenever there's a state change (new block, network /address change)
+
+    if (!isLoading && networkId && userAddress) {
```

Comment:
Yeah, I think so. At least the first call to update will fail until the current fetch chain is complete.

Hmm, how best to abort a recursion in progress?

Because I think if I just ignore it when changing network/address it might add both set of orders to the same list...

Reply:
Well, even without changing anything in the code this is already a problem.

To reproduce:
1. Using metamask, create two accounts and fund each with some ETH for gas, create several liquidity orders to force pagination.
1. Click on orders page and before it finishes loading, switch the account.

Result:
Orders from both accounts will be loaded and a bunch of errors will be printed on the console from react due to duplicate keys:

```
react-dom.development.js?7f13:531 Warning: Encountered two children with the same key, `45`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.
```

Trying to fix it...

Reply:
Idea: Instead of storing a flat list of orders, I'll use a mapping of network and address to avoid overwriting the list under these race conditions. Something like:

```typescript
interface OrdersState {
  orders: AuctionElement[]
  offset: number
}

interface OrdersPerNetwork {
  [networkId: number]: { [address: string]: OrdersState }
}
```

This is already sort of inline with what I was thinking for the global state. I'll try to get this working today...

Reply:
So, managed to get this idea working, slightly different that my suggestion above on https://github.com/gnosis/dex-react/pull/489/commits/d3e94327153abb985b4f695b7df8560487b84925

Works as expected now when switching addresses in the middle of an update, although every time the orders page is loaded something is throwing this error:

```
Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.
    in OrdersWidget (created by Context.Consumer)
```

Couldn't figure out which hook is causing that though. 
And still works as intended, just has this error in the middle. Partial updates (from new blocks) do not raise the error.

Reply:
Great. Sound good enough. 
It keeps adding complexity, although I like that we don't expose out that complexity. 

I think that if we need to do something similar to this again, we could explore other options like https://rxjs-dev.firebaseapp.com to be able to "produce" the orders, stop fetching on certain events and so on in a more natural way